From 41f2ac46c2995bf5bf7ce810f362972216928b37 Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Sun, 27 Dec 2020 12:34:20 +0200
Subject: [PATCH] [SW-27979] i2c indirect: add wdata context

Needed for implementation of opcodes which support WDATA.

So far, the only supported opcode is opcode 12 (fw loader) which is
simple and unique in the way it was implemented. The whole area in which
the data was written was "writable" (no "read only" secion) and the
offset in WDATA was used in a opcode specific way.

Newly implemented opcodes will have to be handled differently, first
it's implementation needs to consider "read only" areas, in which data
may not be written by WDATA (data ignored or raise an error in status).
Secondly WDATA's offset field must be considered.

In order to prepare for new opcode, this commit creates WDATA context
(similar to RDATA context).

Change-Id: I382350e821a799973edd9d40871532b1fd5a58c6
Signed-off-by: Guy Nisan <gnisan@habana.ai>
---

diff --git a/preboot/i2cslave_indirect.c b/preboot/i2cslave_indirect.c
index ae2702d..b21a64e 100644
--- a/preboot/i2cslave_indirect.c
+++ b/preboot/i2cslave_indirect.c
@@ -28,12 +28,17 @@
 	struct hl_i2ci_rdata_hdr hdr;
 };
 
-struct i2cslave_indirect_req {
+struct i2csi_wdata_write_context {
+	u16 length;
 	struct hl_i2ci_wdata_hdr hdr;
+};
+
+struct i2cslave_indirect_req {
 	struct hl_i2ci_req_cfg cfg;
 	struct hl_i2ci_req_sts sts;
 	struct i2csi_wdata_chunk_context chunk;
 	struct i2csi_wdata_read_context rd;
+	struct i2csi_wdata_write_context wd;
 };
 
 static struct i2cslave_indirect_req i2csi_req;
@@ -52,9 +57,10 @@
 	 * of the firmware but rather integrity check data, so we do not
 	 * consider them.
 	 */
-	unsigned int fw_len = i2csi_req.hdr.rqdl +
-		(i2csi_req.hdr.offset * HL_I2CI_REQ_MAX_RQDL) -
-		((i2csi_req.hdr.offset + 1) *  i2csi_req.hdr.ic * 4);
+	unsigned int fw_len = i2csi_req.wd.hdr.rqdl +
+		(i2csi_req.wd.hdr.offset * HL_I2CI_REQ_MAX_RQDL) -
+		((i2csi_req.wd.hdr.offset + 1) *
+		 i2csi_req.wd.hdr.ic * 4);
 
 	if (fit_validate_size(fw_len))
 		return; /* More data is needed */
@@ -66,9 +72,9 @@
 
 }
 
-static inline u8 *i2csi_req_opcode_alloc_buffer(void)
+static inline u8 *i2csi_req_opcode_get_buffer(u16 size)
 {
-	switch (i2csi_req.hdr.opcode) {
+	switch (i2csi_req.wd.hdr.opcode) {
 	case HL_I2CI_REQ_OP_FW_LOADER:
 		/*
 		 * The firmware data is being placed directly to FIT base
@@ -78,23 +84,25 @@
 		 * firmware but rather integrity check data, so we ignore and
 		 * override them.
 		 */
-		return (u8 *)FIT_IMG_BASE + (i2csi_req.hdr.offset *
-			(HL_I2CI_REQ_MAX_RQDL - i2csi_req.hdr.ic * 4));
+		i2csi_req.wd.length = size;
+		return (u8 *)FIT_IMG_BASE + (i2csi_req.wd.hdr.offset *
+			(HL_I2CI_REQ_MAX_RQDL - i2csi_req.wd.hdr.ic * 4));
 	default:
 		/* Data will be consumed from HW but not stored anywhere */
+		i2csi_req.wd.length = size;
 		return NULL;
 	}
 }
 
 static inline void i2csi_req_process_read_opcode(void)
 {
-	switch (i2csi_req.hdr.opcode) {
+	switch (i2csi_req.wd.hdr.opcode) {
 	case HL_I2CI_REQ_OP_DEVICE_STATUS:
 		i2csi_req.rd.data = (u8 *)&op_status;
 		i2csi_req.rd.length = sizeof(op_status);
 		break;
 	default:
-		if ((BIT(i2csi_req.hdr.opcode) & SUPPORTED_OPCODES_MASK))
+		if ((BIT(i2csi_req.wd.hdr.opcode) & SUPPORTED_OPCODES_MASK))
 			i2csi_req.sts.err = HL_I2CI_REQ_ERR_P_WRITE_PERM;
 		else
 			i2csi_req.sts.err = HL_I2CI_REQ_ERR_P_OPCODE;
@@ -104,12 +112,12 @@
 
 static inline void i2csi_req_process_write_opcode(void)
 {
-	switch (i2csi_req.hdr.opcode) {
+	switch (i2csi_req.wd.hdr.opcode) {
 	case HL_I2CI_REQ_OP_FW_LOADER:
 		i2csi_req_load_fw();
 		break;
 	default:
-		if ((BIT(i2csi_req.hdr.opcode) & SUPPORTED_OPCODES_MASK))
+		if ((BIT(i2csi_req.wd.hdr.opcode) & SUPPORTED_OPCODES_MASK))
 			i2csi_req.sts.err = HL_I2CI_REQ_ERR_P_WRITE_PERM;
 		else
 			i2csi_req.sts.err = HL_I2CI_REQ_ERR_P_OPCODE;
@@ -122,20 +130,20 @@
 	register u8 *buf;
 
 	if (cnt == 2) { /* Chunk (partial payload) is expected */
-		buf = i2csi_req_opcode_alloc_buffer();
+		buf = i2csi_req_opcode_get_buffer(i2csi_req.chunk.hdr.rqcl);
 		if (buf)
 			buf += i2csi_req.chunk.offset;
 
 		*data = buf;
 
-		return i2csi_req.chunk.hdr.rqcl;
+		return i2csi_req.wd.length;
 	}
 
 	/* Chuck received */
 	i2csi_req.chunk.offset += i2csi_req.chunk.hdr.rqcl;
 
 	/* Payload fully received */
-	if (i2csi_req.chunk.offset == i2csi_req.hdr.rqdl) {
+	if (i2csi_req.chunk.offset == i2csi_req.wd.hdr.rqdl) {
 		i2csi_req.chunk.offset = 0;
 		i2csi_req.sts.sm_sts = HL_I2CI_REQ_ST_PROCESS;
 		i2csi_req_process_write_opcode();
@@ -148,8 +156,8 @@
 static inline u32 i2csi_req_wdata_handle_payload(u8 **data, u8 cnt)
 {
 	if (cnt == 1) { /* Payload is expected */
-		*data = i2csi_req_opcode_alloc_buffer();
-		return i2csi_req.hdr.rqdl;
+		*data = i2csi_req_opcode_get_buffer(i2csi_req.wd.hdr.rqdl);
+		return i2csi_req.wd.length;
 	}
 
 	/* Payload received */
@@ -192,10 +200,10 @@
 		i2csi_req.chunk.hdr.rqcl = 0;
 		i2csi_req.rd.data = NULL;
 		i2csi_req.rd.length = 0;
-		*data = i2csi_req.hdr.data;
+		*data = i2csi_req.wd.hdr.data;
 		return HL_I2CI_REQ_WDATA_LEN;
 	case 1: /* Chunk header or payload are expected */
-		if (HL_I2CI_REQ_RQCL_IS_EXISTS(i2csi_req.hdr)) {
+		if (HL_I2CI_REQ_RQCL_IS_EXISTS(i2csi_req.wd.hdr)) {
 			*data = i2csi_req.chunk.hdr.data;
 			return HL_I2CI_REQ_CHUNK_LEN;
 		}
@@ -204,7 +212,7 @@
 		break;
 	}
 
-	if (i2csi_req.hdr.write) {
+	if (i2csi_req.wd.hdr.write) {
 		if (i2csi_req.chunk.hdr.rqcl)
 			return i2csi_req_wdata_handle_chunk(data, cnt);
 		else
@@ -219,10 +227,10 @@
 			i2csi_req_process_read_opcode();
 
 		if (i2csi_req.rd.data) {
-			if (i2csi_req.hdr.offset < i2csi_req.rd.length) {
+			if (i2csi_req.wd.hdr.offset < i2csi_req.rd.length) {
 				/* Adjust the read offset */
-				i2csi_req.rd.data += i2csi_req.hdr.offset;
-				i2csi_req.rd.length -= i2csi_req.hdr.offset;
+				i2csi_req.rd.data += i2csi_req.wd.hdr.offset;
+				i2csi_req.rd.length -= i2csi_req.wd.hdr.offset;
 			} else {
 				/* Offset overshoots the read data */
 				i2csi_req.rd.data = NULL;
@@ -230,10 +238,10 @@
 			}
 
 			/* In case master requested partial data */
-			if (i2csi_req.hdr.rqdl < i2csi_req.rd.length)
-				i2csi_req.rd.length = i2csi_req.hdr.rqdl;
+			if (i2csi_req.wd.hdr.rqdl < i2csi_req.rd.length)
+				i2csi_req.rd.length = i2csi_req.wd.hdr.rqdl;
 
-			i2csi_req.rd.hdr.rsdl = i2csi_req.hdr.rqdl;
+			i2csi_req.rd.hdr.rsdl = i2csi_req.wd.hdr.rqdl;
 			i2csi_req.sts.sm_sts = HL_I2CI_REQ_ST_TRANSMIT;
 		} else {
 			i2csi_req.sts.sm_sts = HL_I2CI_REQ_ST_IDLE;
@@ -254,9 +262,9 @@
 	if (idx >= HL_I2CI_REQ_STS_LEN)
 		return 0;
 
-	i2csi_req.sts.opcode = i2csi_req.hdr.opcode;
-	i2csi_req.sts.write = i2csi_req.hdr.write;
-	i2csi_req.sts.ic = i2csi_req.hdr.ic;
+	i2csi_req.sts.opcode = i2csi_req.wd.hdr.opcode;
+	i2csi_req.sts.write = i2csi_req.wd.hdr.write;
+	i2csi_req.sts.ic = i2csi_req.wd.hdr.ic;
 
 	return i2csi_req.sts.data[idx];
 }
